src/Compilers/CSharp/Portable/Symbols/Source/SourceLabelSymbol.cs

```diff
@@ -17,21 +17,48 @@ internal sealed class SourceLabelSymbol : LabelSymbol
         /// </summary>
         private readonly ConstantValue _switchCaseLabelConstant;
 
+        // PERF: Often we do not need this, so we make this lazy
+        private string _lazyName;
+
         public SourceLabelSymbol(
             MethodSymbol containingMethod,
             SyntaxNodeOrToken identifierNodeOrToken,
             ConstantValue switchCaseLabelConstant = null)
-            : base(identifierNodeOrToken.IsToken ? identifierNodeOrToken.AsToken().ValueText : identifierNodeOrToken.ToString())
         {
             _containingMethod = containingMethod;
             _identifierNodeOrToken = identifierNodeOrToken;
             _switchCaseLabelConstant = switchCaseLabelConstant;
         }
 
+        public override string Name
+        {
+            get
+            {
+                return _lazyName ??
+                    (_lazyName = MakeLabelName());
+            }
+        }
+
+        private string MakeLabelName()
+        {
+            var node = _identifierNodeOrToken.AsNode();
+            if (node != null)
+            {
+                return node.ToString();
+            }
+
+            var tk = _identifierNodeOrToken.AsToken();
+            if (tk.Kind() != SyntaxKind.None)
+            {
+                return tk.ValueText;
+            }
+
+            return _switchCaseLabelConstant?.ToString() ?? "";
```

Comment:
I do not think there is a behavior change. We have two mutually exclusive ctors eagerly allocating this string. Now I have one lazy place that allocates it, if needed (that is rare).

The only difference is that I've added " ?? ""  " since switchCaseLabelConstant  is an optional argument defaulting to null. The old code either had hidden NRE or the parameter should not be optional.

I will make the parameter not optional and remove ??.

We have tests for this - all kind of visualizations. They were failing when I had deviations in this code.